### PR TITLE
DBZ-2957 Handle message.key.columns regex validation time complexity

### PR DESCRIPTION
**Pull request by ani-sha**
https://issues.redhat.com/browse/DBZ-2957

### CODE REVIEW

**Comment by gunnarmorling** [comment]
Some comments:

* Let's add a test
* Let's pre-compile the expression into to a `Pattern` instance which is re-used

**Comment by jpechane** [comment]
@ani-sha Have you seen https://issues.redhat.com/browse/DBZ-2957?focusedCommentId=15878234&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-15878234 ?

I think the test coverage needs to be updated for unit tests.

**Comment by ani-sha** [comment]
@jpechane I saw the comments that have been addressed there, but I have a doubt regarding the same. When it was the case of trailing semicolons for `message.key.columns` it kept running and reached a timeout exception so that could be handled by removing the semicolons, but how exactly would it be to handle the random chars at the end of the field or any invalid table names. 
For example, the case that has been provided in the [Jira](https://issues.redhat.com/browse/DBZ-2957?focusedCommentId=15875255&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-15875255):
```
dbo.RbsUsers:integrationRecordId;dbo.RbsUserPrefs:integrationRecordId;dbo.RbsUserCompany:integrationRecordId;dbo.RbsCompanies:integrationRecordId;dbo.RbsCompanyValues:integrationRecordId;dbo.RbsHistory:integrationRecordId;dbo.RbsPassPols:integrationRecordId;dbo.SSOUsers:integrationRecordId;dbo.SSODirectLoginConfig:integrationRecordId;dbo.SSOCoDefaultsHeader:integrationRecordId;asdf
```
The random chars `asdf` would make the field throw an error stating `invalid regex` which is valid IMO. What I mean to say is the value that would be provided is user-centric and we could only validate the right syntax in which it is provided. Isn't it?

**Comment by jpechane** [comment]
@ani-sha I've re-read the original issue and I believe I know better understand the reproters concerns. If you have pretty long configuration and you nake a mistake then it is pretty hard to finda out what went wrong.
So then it really makes sense to change the validation to a different approach

* Split the field value at `;` char
* If the value is empty (two semicolons or semicole at the end) then ignore it
* For every value apply simplified pattern matching - if it fails emit the error for that part that failed

Do you think it is feasible?

**Comment by gunnarmorling** [comment]
To clarify, there still would be a regular expression, but it'd only apply to the individual table-level configuration parts (i.e. it'll be applied after splitting at ";".

**Comment by gunnarmorling** [comment]
Hey @ani-sha, any help needed here?

**Comment by ani-sha** [comment]
@gunnarmorling waiting for a review. I have added the validation check as per your suggestion.

File: debezium-core/src/main/java/io/debezium/relational/Key.java [hunk 1]
```diff
@@ -103,6 +103,8 @@ public static KeyMapper getInstance(String fullyQualifiedColumnNames, TableIdToS
                 return null;
             }
 
+            fullyQualifiedColumnNames = fullyQualifiedColumnNames.replaceAll(";+$", "");
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/Key.java:106]: Is this still needed any longer as per your latest changes?

File: debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java [hunk 2]
```diff
@@ -712,9 +712,16 @@ private static int validateDatabaseExcludeList(Configuration config, Field field
 
     private static int validateMessageKeyColumnsField(Configuration config, Field field, Field.ValidationOutput problems) {
         String msgKeyColumns = config.getString(MSG_KEY_COLUMNS);
-        if (msgKeyColumns != null && !MSG_KEY_COLUMNS_PATTERN.asPredicate().test(msgKeyColumns)) {
-            problems.accept(MSG_KEY_COLUMNS, msgKeyColumns, MSG_KEY_COLUMNS.name() + " has an invalid format (expecting '" + MSG_KEY_COLUMNS_PATTERN.pattern() + "')");
-            return 1;
+        if (msgKeyColumns != null) {
+            if (msgKeyColumns.contains(";")) {
+                msgKeyColumns = msgKeyColumns.replaceAll(";+$", "");
+            }
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java:718]: Is this one actually needed with the split you do right below?

File: debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java [hunk 3]
```diff
@@ -712,9 +712,16 @@ private static int validateDatabaseExcludeList(Configuration config, Field field
...
+            for (String substring : msgKeyColumns.split(";")) {
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java:719]: Let's extract a pattern instance and re-use it:

    private static final Pattern COLUMN_SPLIT = Pattern.compile(";");

And here: `COLUMN_SPLIT.split(msgKeyColumns) ...`.

File: debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java [hunk 4]
```diff
@@ -48,7 +48,7 @@
     protected static final String TABLE_EXCLUDE_LIST_NAME = "table.exclude.list";
     protected static final String TABLE_WHITELIST_NAME = "table.whitelist";
     protected static final String TABLE_INCLUDE_LIST_NAME = "table.include.list";
-    private static final Pattern MSG_KEY_COLUMNS_PATTERN = Pattern.compile("^(([^:]+):([^:;\\s]+))+[^;]$");
+    private static final Pattern MSG_KEY_COLUMNS_PATTERN = Pattern.compile("^(([^:]+):([^:;\\s]+))");
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java:51]: Should it be MSG_KEY_COLUMN_PATTERN now (no "S")?

File: debezium-core/src/test/java/io/debezium/config/ConfigurationTest.java [hunk 5]
```diff
@@ -254,4 +254,17 @@ public void testMsgKeyColumnsField() {
         config = Configuration.create().with(MSG_KEY_COLUMNS, "t1,t2").build();
         assertThat(config.validate(Field.setOf(MSG_KEY_COLUMNS)).get(MSG_KEY_COLUMNS.name()).errorMessages()).isNotEmpty();
     }
+
+    @Test
+    @FixFor("DBZ-2957")
+    public void testMsgKeyColumnsFieldRegexValidation() {
+        config = Configuration.create().with(MSG_KEY_COLUMNS, "t1:C1;(.*).t2:C1,C2;").build();
+        assertThat(config.validate(Field.setOf(MSG_KEY_COLUMNS)).get(MSG_KEY_COLUMNS.name()).errorMessages()).isEmpty();
```
Comment (gunnarmorling) [inline debezium-core/src/test/java/io/debezium/config/ConfigurationTest.java:262]: Perhaps assert the actual message(s)? I.e. which one of of the columns isn't right? That said, I think we still should be lenient towards a trailing semicolon, we can properly interpret the expression also with that. Any other trailing part (e.g. "t1:C1;foobar") would be flagged by your new logic already.

**Review by gunnarmorling — COMMENTED** [review]
A few comments inline. We're getting there :)

**Comment by ani-sha** [comment]
@gunnarmorling addressed the changes.

File: debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java [hunk 6]
```diff
@@ -48,7 +48,8 @@
     protected static final String TABLE_EXCLUDE_LIST_NAME = "table.exclude.list";
     protected static final String TABLE_WHITELIST_NAME = "table.whitelist";
     protected static final String TABLE_INCLUDE_LIST_NAME = "table.include.list";
-    private static final Pattern MSG_KEY_COLUMNS_PATTERN = Pattern.compile("^(([^:]+):([^:;\\s]+))+[^;]$");
+    private static final Pattern MSG_KEY_COLUMNS_PATTERN = Pattern.compile("^(([^:;]+):([^:;\\s]+))+$");
```
Comment (gunnarmorling) [inline debezium-core/src/main/java/io/debezium/relational/RelationalDatabaseConnectorConfig.java:51]: Now that we already split up the expression at ";" before, this expression can be simplified a bit. In particular, the ";" can be removed from the negated classes.

**Review by gunnarmorling — COMMENTED** [review]
@ani-sha, ok, I think it's almost good to go. One last suggestion inline.

**Comment by gunnarmorling** [comment]
In fact, I'll quickly do that last change myself and update this PR; thanks, @ani-sha!

**Comment by gunnarmorling** [comment]
@ani-sha, I've pushed two commits to adjust a few things. I suggest you take a look so to see whether it's ok for you. Thx!

**Comment by gunnarmorling** [comment]
Gnarf, did escaping wrong at first. Will squash my commits into one eventually.

**Comment by gunnarmorling** [comment]
Humpf, seems my changed caused a regression for Vitess. Will need to take a look into it tomorrow, unless you beat me to it, @ani-sha :)

**Comment by gunnarmorling** [comment]
I've created https://github.com/debezium/debezium/pull/2230 for continuing the work, closing this one. Thanks a lot, @ani-sha!

**Pull request closed without merging**